metrics_reporter.go

```diff
@@ -0,0 +1,125 @@
+package shuttle
+
+import (
+	"fmt"
+	"time"
+	"log"
+
+	"github.com/heroku/slog"
+	metrics "github.com/rcrowley/go-metrics"
+)
+
+// Percentile info for histograms and the like. These 2 arrays need to match
+// length and positions are relevant to each other.
+var (
+	percentiles     = []float64{0.75, 0.95, 0.99}
+	percentileNames = []string{"p75", "p95", "p99"}
+)
+
+// Given a t representing a time in ns, convert to seconds, show up to μs precision
+func sec(t float64) string {
+	return fmt.Sprintf("%.6f", t/1000000000)
+}
+
+// MetricsReporter handles reporting of metrics to a specified source at a given duration
+type MetricsReporter struct {
+	registry metrics.Registry
+	source   string
+	duration time.Duration
+	logger   *log.Logger
+	lastCounts map[string]int64
+	doneCh chan struct{}
+}
+
+// NewMetricsReporter returns a properly constructed MetricsReporter
+func NewMetricsReporter(r metrics.Registry, source string, d time.Duration, l *log.Logger) *MetricsReporter {
+	return &MetricsReporter{registry: r, source: source, duration: d, logger: l, lastCounts: make(map[string]int64), doneCh: make(chan struct{})}
+}
+
+func (e MetricsReporter) countDifference(ctx slog.Context, name string, c int64) {
+	name = name + ".count"
+	lc := e.lastCounts[name]
+	ctx[name] = c - lc
+	e.lastCounts[name] = c
+}
+
+// Emit emits log-shuttle metrics in logfmt compatible formats every d
+// duration using the MetricsReporter logger. source is added to the line as
+// log_shuttle_stats_source if not empty. It waits for a stop signal
+// and will stop emitting metrics when received.
+func (e MetricsReporter) Emit() {
+	if e.duration == 0 {
```

Comment:
Observation
Duration is only really needed here, so could be moved away from being a struct member and to an Emit function parameter. This makes more sense to me since personally, but I can understand if it doeesn't to you.